src/librustc_borrowck/borrowck/gather_loans/mod.rs

```diff
@@ -233,7 +233,17 @@ fn check_mutability<'a, 'tcx>(bccx: &BorrowckCtxt<'a, 'tcx>,
                                            cmt,
                                            code: err_mutbl }))
             } else {
-                Ok(())
+                match cmt.maybe_static_mutability() {
```

Comment:
Something about this check doesn't feel right. There is a specific check that is supposed to guarantee that the reference is not an "aliasable" location, and it seems to be going awry, but I feel like we should be fixing *that* check, and not adding a new one. This is the check I am talking about:

https://github.com/rust-lang/rust/blob/5965b790142eff7a8546e947914e7a8e00c61575/src/librustc_borrowck/borrowck/gather_loans/mod.rs#L164-L203

Can you maybe tell me what the `cmt` is in this case, and what the value of `aliasability` is?

Reply:
The `cmt` is
```rust
cmt=cmt_ {
        id: NodeId(16),
        span: test.rs:13:5: 13:11,
        cat: Deref(cmt_ {
            id: NodeId(16),
            span: test.rs:13:5: 13:11,
            cat: Interior(cmt_ {
                id: NodeId(14),
                span: test.rs:13:5: 13:8,
                cat: StaticItem,
                mutbl: McImmutable,
                ty: [&'static mut [u8]; 0],
                note: NoteNone,
            }, []),
            mutbl: McImmutable,
            ty: &'static mut [u8],
            note: NoteNone,
        }, BorrowedPtr(MutBorrow, ReStatic)),
        mutbl: McDeclared,
        ty: [u8],
        note: NoteNone
    }
```

and the value of `aliasability` is
```rust
aliasability=FreelyAliasable(AliasableStatic)
```

Reply:
Huh, so I wonder in that case why this code:

https://github.com/rust-lang/rust/blob/5965b790142eff7a8546e947914e7a8e00c61575/src/librustc_borrowck/borrowck/gather_loans/mod.rs#L190-L198

doesn't execute... `req_kind` is `ty::MutBorrow`, right?

Reply:
That is because `check_mutability` triggers an error, and `check_aliasability` is no called at all.

Reply:
Huh. So how come I don't see an error reporting in the original issue #42344?  I see the ICE -- where does that ICE occur?

Reply:
Oh, were you asking about the behavior of the original rustc? I was talking about how rustc bahaves when this PR was applied. I am sorry for the confusion.

Using the master branch, yes, `req_kind` is `ty::MutBorrow`, and `bccx.report_aliasability_violation` is executed. The following error is reported.

```
error: internal compiler error: aliasability violation for static `cannot borrow data mutably`
 --> test.rs:3:5
  |
3 |     TAB[0].iter_mut();
  |     ^^^^^^

error: aborting due to previous error
```

Reply:
@topecongiro ah, I see -- actually, I think that the `delay_span_bug` is just wrong and should be removed. The logic was that a case like `static X: u32 = 2; &mut X` should not reach there (and indeed I think it won't) -- but it overlooked the possibility of an `&mut` living in a `static`.

Reply:
Can you update this PR to just remove the `delay_span_bug` call for `AliasabilityReason::Static`?

Reply:
Sure! Thank your for your patience with my slow updates.